### ISSUE TITLE
Heat for RACs is too low in v.13

### ISSUE THREAD

**Issue by sestrand** [lisongmechlab/lsml#757]
**LSML version:** 2.0.13
**Screen resolution:** 3440x1440
**Java version:** 8u301

**Steps to reproduce issue:**
1. Open LSML
2. Click on garage
3. Then on mech 
4. then mouse over rac5 or rac2 and get the tooltip for that weapon.

**Actual result:**
heat  per shot for rac5 is reported as 1.59 heat per second and 0.99 heat per second for RAC2. and the heat for the 5sec etc are not nearly high enough.

**Expected result:**
in game (and in the data file) heat is reported as 3.25 heat per second for rac5 and 2 heat per second for RAC2.

the RACs all have a rate of fire of 7.275 rounds per second (per the data file) and that equates to a cooldown of .137457. which places the heat per shot at .4467 for the RAC5 and .2749 for the RAC2

I am not certain where this is slipping but I suspect the use of eFP as the culprit ...  the LSML weapons table reports correctly for rFP and Ht.   I am not certain how eFP plays into the heat calculation, but if you pull and hold the trigger you are getting rFP rate of heat regardless of the eFP.  eFP seems a little low as well post a recent patch, since you can now fire the RAC5 for 8 seconds before there is a chance of jam.  (per the data file and patch notes).  meaning if you can keep your face time to 5 seconds or so, you will never jam with a rac. 

my recommendation is that rFP be used for the heat calculation. and if jamming is being considered, eFP would be only used for damage estimate.  probably true for uacs as well. The  heat should not be underestimated and rFP would err on the side of too high, but only by a little bit.

thanks for your continued support on this tool.

**Comment by LiSongMWO** [comment]
Please take care to be consistent in your report. You say "heat per shot is 1.59 heat per second" which is nonsensical and there is no such thing shown in the tooltip. Do you mean the "Heat" (which is heat per shot) or "Heat per Second" which is a different thing?

As far as I can see the "Heat" field correctly shows 0.45 (which is 0.4467 *correctly* rounded to two significant digits). The "Heat per Second" field shows eHPS which is consistent with the weapons table. However the tooltip can be clarified to say that it's eHPS.

I'm sensing a misunderstanding here:  eFP (and eHPS similarly) is the statistically [expected value](https://en.wikipedia.org/wiki/Expected_value) as used in probability theory. You can think of it like this, you perform an infinite number of experiments, in each of them you have infinite ammo, and hold down the trigger until the heat death of the universe and then take the average of that run, and then you take the average over all the infinite numbers of experiments, the result from all that is eFP. In layman's terms, it means that it's what you should expect to get on average over many, long engagements.

In particular eFP includes jams and jam free time, and during the jams you generate no heat. Therefore it's correct to compute eHPS=Ht/eFP and rHPS=Ht/rFP (not shown in LSML). Computing eHPS=Ht/rFP is incorrect because you're accounting for more time generating heat than what is expected for eHPS. You probably *wanted* to see rHPS = Ht/rFP, but LSML currently doesn't show this anywhere. 

eDPS is significantly lower than rDPS (in other words, eFP is longer than rFP) for RAC due to it's rather hefty jam penalty. But eDPS of RAC is still higher than other IS ballistics of similar calibre (UAC5, LB5X) in order to make up for it's drawbacks of spin-up and required face time. This is the balancing factor of RAC, if eDPS was higher it would be absolutely broken.

I have computed eFP in several two different ways and they both give the same result. I have several automated tests that agree that the results are correct.  There is no evidence to suggest that eFP is incorrect for RAC or UAC after 2.0.13.

If you think eFP is wrong I will need strong evidence: Load up an IS assault mech without ballistics cooldown quirks with two RAC5s, enough double heat sinks to never overheat and fill the rest with ammo like [this](http://t.li-soft.org/?l=rwJsQEoGTxJuCkoGT0CSpSnGUISkB9faGtDRoa0NaGtDWhrQ1oa0NaGjQ1oaNDR81oa0NGhrQ1oa0NaGtDWhrQ1oaNDWho0NaGvr5g%3D%3D). Then go into a lobby game, and time how long it takes to empty all the ammo using a stop watch (should be around 12 minutes). Divide the time by total number of shots you had, that's the average FP (not eFP!!) for that experiment, then repeat that experiment around 14 times for statistical significance, take the average of those and if that's significantly different than eFP reported by LSML then there is reason to believe the eFP calculation could be wrong.

Using rFP for eHPS would not just be "a little bit too hot" it would be almost twice as hot for RACs. And using rFP for eHPS for UACs is again too low because of the double fire not being accounted for. It would be wrong both ways for what it is computed, eHPS. Choosing clear metrics that work for everything is hard, LSML's metrics are consistent as long as you are aware of what they actually are. Again, freely admitting there are improvements that can be made on the clarity front.

What I'm suspect you're actually wanting is: a"peak dps" and "peak heat" metric to show you what you get during maximum luck with RAC or UAC, including double fire of UAC (note that this is not rFP, this is a new value minFP which is the minimum possible average firing period. For RAC rFP=minFP but for UAC rFP > minFP due to double fire). As I'm only maintaining LSML to prevent it from falling into disrepair, I'm not going to implement these anytime soon. However if someone wants to contribute the code, I'm happy to merge it. I agree the tooltip wording could be updated to be more clear.

That said, I have already spent way more time patching LSML these past weeks than I wanted to and I'm exhausted and I want some of my free time back. As I said, I have no reason to believe that either eHPS or eFP is incorrect (they may not be what you wanted them to be, but I have no reason to believe they're broken from what they're intended to be). Neither do I have any reason to believe that the time and heat values shown in the right hand side panel are wrong. They may not be super clearly labeled that they are all expected values (except for burst and alpha). That's something that can be improved in a future release but unless something is severely broken, I want to not have to touch LSML for a hot minute.

Because this issue seems to collate many different aspects, I ask that you split this into several issues for updating tooltips, labeling of the metrics and adding the additional metrics if those are what you want. Otherwise it's hard to track work on these separately.

Again I want to stress: I'm only fixing bugs and minor UI things. No new development from my part, happy to accept contributions.

**Comment by sestrand** [comment]
sorry about the mis-type.  yes heat per second is in the tool tip.

I think you have understood my concern.  I did figure out what this
is intended to represent, that this is a burst that could happen at any
time (not from 0 temp and rac cycle).  peak heat is of more interest to me
than peak damage. but having both displayed would be helpful.  peak heat is
what you encounter in the game in those first 8 seconds for RAC5.  so
understanding what that does to your mech is helpful. having a way to
display it would be helpful.  and yes it is twice as hot as the eHt being
reported for RAC2s.   I just would like to see the peak in the display to
account for that initial volley (or at any point that you can get them back
to zero before starting it up again) where there is no chance of jam.

this additional view is more important for RAC than UAC because of the way
jamming differs.  currently RACs have a long enough grace period that they
will jam about the time a mech is hitting 100% heat.  UACs could go out on
the first volley.  I do not think that the way you calculate eHt is wrong
for RACs but it doesn't paint the full picture.

I also very much understand the time limits.  I will try to capture each
panel in an individual issue.  thanks again for your updates.  I'll see
what I can do to contribute. (when I have some time as well)

**Comment by LiSongMWO** [comment]
Given your last comment about opening new issues. Are we ok to close this one?

**Comment by LiSongMWO** [comment]
I'll take that as a yes.
